### ISSUE TITLE
Add a directive that allows the fetching of tool version meta information

### ISSUE THREAD

**Issue by pditommaso** [nextflow-io/nextflow#879]
The goal of this proposal is to add a process directive that allow the user to specify one-liner command to be executed as very first execution to fetch the tool command version. For example: 

```nextflow
process foo {
  container: {   }
   versionCmd: "bwa --version"

   input: 
   file x from z
   script:
   """
   do-this.sh
   """
}
```

The version command should be executed as the specified by the process configuration i.e. using the provided executor/container/module settings. 

More than one `versionCmd` should be possible. 

To avoid to spawn multiple jobs the `versionCmd` command should be prefixed to the real command script execution, only the first time. Moreover the version command should be trigger each time a dynamic container or module configuration is applied.

**Comment by pditommaso** [comment]
This may required #540

**Comment by pditommaso** [comment]
Also this version should be collected to include in the execution report and trace report. As possible extension it should be possible to collect these information using a channel and send to a *reporting* process (reads MultiQC).

**Comment by pditommaso** [comment]
Brainstormed a bit more with @emi80 about this. Since each task can potentially use a different container or run in a different host with a different env (when container are not used). 

The proper way to handle this would be to executed the `versionCmd` *before* each task execution and capture the output in the command log using a special marker. Then NF should parse this log fetch the version info and track in the task execution metadata. 

The `versionCmd` can be an arbitrary bash script that can chain one more command to properly dump the version of one or more tools used in the target task script.

**Comment by ewels** [comment]
Yes, I guess having it on task level is better. Shouldn’t be too tricky to collapse duplicates for MultiQC etc..

**Comment by pditommaso** [comment]
the new pic! 😆

**Comment by ewels** [comment]
Bump

**Comment by pditommaso** [comment]
(sorry for the spam! 😄)

**Comment by ewels** [comment]
No problem! I'm not bumping all of these, I'm letting the ones I don't care about much die quietly 😅

**Comment by bentsherman** [comment]
Solved by eval output type #4493
